Handle variable expansion for paths and managed env names for includes

kwryankrattiger opened the pull request:
Allow including environments using the managed env name or using config path alias like `$spack` or environment variables (as can be used with `include` paths).

Also update the CLI to not expand the path when adding included concrete environments to the config. This should allow better consistency between what is expressed on the CLI and what is reflected in the config.

kwryankrattiger committed:
Expand paths and managed env name for included concrete

kwryankrattiger committed:
Expand paths and managed env name for included concrete

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update unit test to reflect new behavior

kwryankrattiger committed:
Make error messages unique to track environment errors

kwryankrattiger committed:
Make error messages unique to track environment errors

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

tldahlgren commented:
> Allow including environments using the managed repo name or using config path alias like `$spack` or enironment variables like `include` paths.

I can see the value of this.

> Also update the CLI to not expand the path when adding included concrete environments to the config. This should allow better consistency between what is expressed on the CLI and what is reflected in the config.

I'm not so sure about this change. I thought it was a deliberate choice to use expanded managed environment paths for concrete environments.  

@tgamblin @RikkiButler20 What are your thoughts on this?

tldahlgren commented:
IMO There needs to be at least one unit test that relies on variable substitution for an included concrete environment.

kwryankrattiger commented:
The reason for the CLI change is I don't like my configs being different than what I have specified. If I say "use the managed name", then I would expect spack to do that in the config. If I specify the full path then spack should do that, not collapse it to the managed name.

The benefit of using the name is no matter where my environments exist, the path is always computed from the current state. Not having that makes my config more difficult to relocate by default, which in many of my workflows is problematic as I move and reuse configs liberally.

I would be interested to hear the argument for always expanding the managed environment name to a path for including concrete. I didn't see anything in the code that described it as a feature nor documentation about the behavior other than that is what the current code does.

kwryankrattiger committed:
Use variable expantion in include_concrete test

kwryankrattiger committed:
Use variable expantion in include_concrete test

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger commented:
> I'm not so sure about this change. I thought it was a deliberate choice to use expanded managed environment paths for concrete environments.

Note, we do expand all of the paths of included concrete environments for all of the internal usage. The _only_ place this is adding support for the un-expanded paths is in `spack.yaml`. As soon as they are read, they are immediately expanded and referenced and cached using the full expanded path. This makes sense for normalization and is consistent with how we handle other path configs.

kwryankrattiger committed:
Updates post rebase

Some new things have been added/changed already to better support this
but there seem to still be some places missing correct expansion.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Fix another place missing expansion for concrete

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>
